src/request_executor.cc

```diff
@@ -245,98 +346,29 @@ RequestExecutor::Infer(
     }
 
     {
-      auto p = new std::promise<TRITONSERVER_InferenceResponse*>();
-      std::future<TRITONSERVER_InferenceResponse*> completed = p->get_future();
+      auto p = new std::promise<std::unique_ptr<InferResponse>>();
+      response_future = p->get_future();
+      infer_request->SetPrevPromise(std::move(&p));
```

Comment:
Why do we need to attach the promise to the inference request? How about just passing the response_future as the `userp` in the callback?

Reply:
Sorry I'm not sure what you mean here. The `response_future` here lets us to retrieve the response later when we need to.  Would it be possible if we return the `response_future` and pass it to the callback function as well?
We need to pass the promise to the callback function so that we can set the current `InferResponse` to the promise. We also need to check if the request is for decoupled model or not, and the information can be retrieved by `infer_request->IsDecoupled()`. The promise is stored inside the `inferRequest` object so that the lifetime of the promise would be long enough for  calling `set_value` later with the next response.

Reply:
Discussed offline.